core/controllers/admin_test.py

```diff
@@ -1109,18 +1111,154 @@ def test_send_dummy_mail(self):
+            '/updateusernamehandler',
+            payload={
+                'old_username': 123,
+                'new_username': self.NEW_USERNAME},
+            csrf_token=csrf_token,
+            expected_status_int=400)
+        self.assertEqual(
+            response['error'], 'Expected old username to be a unicode '
+            'string, received 123')
+
+    def test_update_username_with_long_new_username(self):
+        long_username = 'a' * (constants.MAX_USERNAME_LENGTH + 1)
+        csrf_token = self.get_new_csrf_token()
+
+        response = self.put_json(
+            '/updateusernamehandler',
+            payload={
+                'old_username': self.OLD_USERNAME,
+                'new_username': long_username},
+            csrf_token=csrf_token,
+            expected_status_int=400)
+        self.assertEqual(
+            response['error'], 'Expected new username to be less than %s '
+            'characters, received %s' % (
+                constants.MAX_USERNAME_LENGTH,
+                long_username))
+
+    def test_update_username_with_nonexistent_old_username(self):
+        non_existent_username = 'invalid'
+        csrf_token = self.get_new_csrf_token()
+
+        response = self.put_json(
+            '/updateusernamehandler',
+            payload={
+                'old_username': non_existent_username,
+                'new_username': self.NEW_USERNAME},
+            csrf_token=csrf_token,
+            expected_status_int=400)
+        self.assertEqual(response['error'], 'Invalid username: invalid')
+
+    def test_update_username_with_new_username_already_taken(self):
+        csrf_token = self.get_new_csrf_token()
+
+        response = self.put_json(
+            '/updateusernamehandler',
+            payload={
+                'old_username': self.OLD_USERNAME,
+                'new_username': self.OLD_USERNAME},
+            csrf_token=csrf_token,
+            expected_status_int=400)
+        self.assertEqual(response['error'], 'Username already taken.')
+
+    def test_update_username(self):
+        user_id = self.get_user_id_from_email(self.ADMIN_EMAIL)
+        csrf_token = self.get_new_csrf_token()
+
+        self.put_json(
+            '/updateusernamehandler',
+            payload={
+                'old_username': self.OLD_USERNAME,
+                'new_username': self.NEW_USERNAME},
+            csrf_token=csrf_token)
+        self.assertEqual(user_services.get_username(user_id), self.NEW_USERNAME)
+
+    def test_update_username_creates_audit_model(self):
+        user_id = self.get_user_id_from_email(self.ADMIN_EMAIL)
+        csrf_token = self.get_new_csrf_token()
+
+        self.put_json(
+            '/updateusernamehandler',
+            payload={
+                'old_username': self.OLD_USERNAME,
+                'new_username': self.NEW_USERNAME},
+            csrf_token=csrf_token)
+
+        self.assertTrue(
+            audit_models.UsernameChangeAuditModel.has_reference_to_user_id(
```

Comment:
Add check to validate fields of the audit model.

Reply:
Done!